keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1811 @@
+we will need a separate e2e job to test the nftables mode (like we do
+with ipvs). Eventually, if nftables becomes the default, then this
+would be flipped around to having a legacy "iptables" job.
+
+The test "`[It should recreate its iptables rules if they are
+deleted]`" tests (a) that kubelet recreates `KUBE-IPTABLES-HINT` if it
+is deleted, and (b) that deleting all `KUBE-*` iptables rules does not
+cause services to be broken forever. The latter part is obviously a
+no-op under `nftables` kube-proxy, but we can run it anyway. (We are
+currently assuming that we will not need an nftables version of this
+test, since the problem of one component deleting another component's
+rules should not exist with nftables.)
+
+(Though not directly related to kube-proxy, there are also other e2e
+tests that use iptables which should eventually be ported to nftables;
+notably, the ones using [`TestUnderTemporaryNetworkFailure`].)
+
+For the most part, we should not need to add any nftables-specific e2e
+tests; the `nftables` backend's job is just to implement the Service
+proxy API to the same specifications as the other backends do, so the
+existing e2e tests already cover everything relevant. The only
+exception to this is in cases where we change default behavior from
+the `iptables` backend, in which case we may need new tests for the
+different behavior.
+
+We will eventually need e2e tests for switching between `iptables` and
+`nftables` mode in an existing cluster.
+
+[It should recreate its iptables rules if they are deleted]: https://github.com/kubernetes/kubernetes/blob/v1.27.0/test/e2e/network/networking.go#L550
+[`TestUnderTemporaryNetworkFailure`]: https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/framework/network/utils.go#L1078
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+#### Scalability & Performance tests
+
+```
+<<[UNRESOLVED perfscale ]>>
+
+- For the control plane side, the existing scalability tests are
+  probably reasonable, assuming we implement the same
+  `NetworkProgrammingLatency` metric as the existing backends.
+
+- For the data plane side, there are tests of
+  `InClusterNetworkLatency`, but no one is really looking at the
+  results yet and they may need work before they are useable.
+
+- We should also make sure that other metrics (CPU, RAM, I/O, etc)
+  remain reasonable in an `nftables` cluster.
+
+<<[/UNRESOLVED]>>
+```
+
+### Graduation Criteria
+
+#### Alpha
+
+- `kube-proxy --proxy-mode nftables` available behind a feature gate
+
+- nftables mode has unit test parity with iptables
```

Comment:
What is the test coverage of iptables?

Playing devils advocate a bit, if it would be 10%, then it's not good enough criteria.
Can you please fill in the test coverage for iptables/ipvs (there are links in KEP template) and then potentially adjust here if needed?

Reply:
(https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit does not seem to be loading for me? I get the header but then it just says "Loading results from server..." forever. Or does it just take a really really really long time?)

But from KEP-3453:

```
- `k8s.io/kubernetes/pkg/proxy/iptables`: `2022-08-02` - `67%`
```

and it's probably still within 10% of that

Reply:
from current master
```
coverage: 70.6% of statements
ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.364s	coverage: 70.6% of statements
```

Reply:
@danwinship - can you please paste the above to the KEP?

Now, getting back to the core of my question - I understand that we generally won't be touching iptables backend so requiring changes there would be counter-productive.

But given it will be replacement for the existing things, and as identifier by @liggitt those are things where we observe the biggest number of regressions, I think that our bar for coverage for nftables should actually be higher.

WDYT?

Reply:
I already did add the iptables number to [the "Unit tests" section](https://github.com/kubernetes/enhancements/pull/3824/files#diff-85ab4fa1a983d0d1d33b6bc089dc2c5c0b558d73672e7e2afad8a647ad210061R1321)

The rest of your comment seems definitely post-alpha?

Reply:
> The rest of your comment seems definitely post-alpha?

For unit tests I would actually expect those coming together with the implementation.

For all other tests, having them post alpha makes sense.